detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessaryNotNullCheck.kt

```diff
@@ -0,0 +1,61 @@
+package io.gitlab.arturbosch.detekt.rules.bugs
+
+import io.gitlab.arturbosch.detekt.api.CodeSmell
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.api.Debt
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Rule
+import io.gitlab.arturbosch.detekt.api.Severity
+import io.gitlab.arturbosch.detekt.api.internal.RequiresTypeResolution
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.psiUtil.getCallNameExpression
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.util.getType
+import org.jetbrains.kotlin.types.typeUtil.TypeNullability
+import org.jetbrains.kotlin.types.typeUtil.nullability
+
+/**
+ * Reports unnecessary not-null checks with `requireNotNull` or `checkNotNull` that can be removed by the user.
+ *
+ * <noncompliant>
+ * var string = "foo"
+ * println(requireNotNull(string))
+ * </noncompliant>
+ *
+ * <compliant>
+ * var string : String? = "foo"
+ * println(requireNotNull(string))
+ * </compliant>
+ */
+@RequiresTypeResolution
+class UnnecessaryNotNullCheck(config: Config = Config.empty) : Rule(config) {
+
+    override val issue = Issue(
+        "UnnecessaryNotNullCheck",
+        Severity.Defect,
+        "Unnecessary not-null check detected.",
```

Comment:
```suggestion
        "Remove unnecessary not-null checks on non-null types.",
```
I would recommend using a more elaborative description since our description has been historically vague.